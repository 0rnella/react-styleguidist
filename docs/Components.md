# Locating your components

## List of components

By default Styleguidist will search component using this glob mask: `src/components/**/*.js`. If it doesn’t work for you, create a **`styleguide.config.js`** file in your project’s root folder.

If your components look like `components/Button.js` or `components/Button/Button.js` or `components/Button/index.js` (it’s the default behavior, change only if your components are not in `src/components` or files are not `.js`):

```javascript
module.exports = {
  components: 'src/components/**/*.js'
};
```

If your components look like `components/Button/Button.js` + `components/Button/index.js` (you need to skip `index.js`, otherwise component will be loaded twice):

```javascript
module.exports = {
  components: 'src/components/**/[A-Z]*.js'
};
```

Or you need to skip test specs (`components/Button/Button.test.js`):

```javascript
module.exports = {
  components: 'components/**/!(*.test).js'
};
```

## Sections

Allows components to be grouped into sections with a title and optional overview content. Sections can also be content only, with no associated components (for example, a textual introduction). A section definition consists of optional fields:

- `name` — the title of the section.
- `content` — location of a Markdown file containing the overview content.
- `components` — a string or function returning a list of components. The same rules apply as for the root `components` option.
- `sections` — array of subsections.

Configuring a style guide with a textual introduction section, then a UI section would look like:

```javascript
module.exports = {
  sections: [
    {
      name: 'Introduction',
      content: 'docs/introduction.md'
    },
    {
      name: 'Documentation',
      sections: [
        {
          name: 'Installation',
          content: 'docs/installation.md'
        },
        {
          name: 'Configuration',
          content: 'docs/configuration.md'
        }
      ]
    },
    {
      name: 'UI Components',
      content: 'docs/ui.md',
      components: 'lib/components/ui/*.js'
    }
  ]
};
```
