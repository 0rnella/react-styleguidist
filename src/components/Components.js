import React, { PropTypes } from 'react';
import Component from './Component';

export default React.createClass({
	displayName: 'Components',
	propTypes: {
		components: PropTypes.array.isRequired
	},

	renderComponents() {
		return this.props.components.map((component) => {
			return (
				<Component component={component} key={component.name}/>
			)
		});
	},

	render() {
		return (
			<div>
				{this.renderComponents()}
			</div>
		);
	}
});
