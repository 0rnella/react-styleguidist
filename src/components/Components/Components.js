import { Component, PropTypes } from 'react';
import ReactComponent from 'components/ReactComponent';

export default class Components extends Component {
	static propTypes = {
		components: PropTypes.array.isRequired
	}

	renderComponents() {
		return this.props.components.map((component) => {
			return (
				<ReactComponent component={component} key={component.name}/>
			)
		});
	}

	render() {
		return (
			<div>
				{this.renderComponents()}
			</div>
		);
	}
}
