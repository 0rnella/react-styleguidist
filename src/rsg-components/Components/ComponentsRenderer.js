import React, { PropTypes } from 'react';

export default function ComponentsRenderer({ components, sections }) {
	return (
		<div>
			{components}
			{sections}
		</div>
	);
}
ComponentsRenderer.propTypes = {
	components: PropTypes.array.isRequired,
	sections: PropTypes.node.isRequired,
};
