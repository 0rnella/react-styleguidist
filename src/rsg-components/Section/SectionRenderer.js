import React from 'react';
import PropTypes from 'prop-types';
import Styled from 'rsg-components/Styled';
import Heading from 'rsg-components/Heading';

const styles = ({ space, fontFamily, fontSize }) => ({
	root: {
		marginBottom: space[4],
	},
	heading: {
		margin: [[0, 0, space[2]]],
		fontFamily: fontFamily.base,
		fontSize: fontSize.h1,
	},
});

export function SectionRenderer({ classes, name, slug, content, components, sections }) {
	return (
		<section className={classes.root}>
			{name && (
				<Heading level={1} slug={slug} className={classes.heading}>{name}</Heading>
			)}
			{content}
			{components}
			{sections}
		</section>
	);
}

SectionRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	name: PropTypes.string,
	slug: PropTypes.string,
	content: PropTypes.node,
	components: PropTypes.node,
	sections: PropTypes.node,
};

export default Styled(styles)(SectionRenderer);
