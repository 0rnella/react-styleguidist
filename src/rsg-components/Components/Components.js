import React, { Component, PropTypes } from 'react';
import ReactComponent from 'rsg-components/ReactComponent';
import Sections from 'rsg-components/Sections';
import ComponentsRenderer from 'rsg-components/Components/ComponentsRenderer';

export default class Components extends Component {
	static propTypes = {
		highlightTheme: PropTypes.string.isRequired,
		components: PropTypes.array.isRequired,
		sections: PropTypes.array.isRequired,
		sidebar: PropTypes.bool,
	};

	renderComponents() {
		const { highlightTheme, components, sidebar } = this.props;

		return components.map(component => {
			return (
				<ReactComponent
					key={component.filepath}
					highlightTheme={highlightTheme}
					component={component}
					sidebar={sidebar}
				/>
			);
		});
	}

	renderSections() {
		const { highlightTheme, sections, sidebar } = this.props;

		return (
			<Sections
				highlightTheme={highlightTheme}
				sections={sections}
				sidebar={sidebar}
			/>
		);
	}

	render() {
		return (
			<ComponentsRenderer
				components={this.renderComponents()}
				sections={this.renderSections()}
			/>
		);
	}
}
