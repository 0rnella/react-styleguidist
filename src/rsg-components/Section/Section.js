import React, { Component, PropTypes } from 'react';
import Examples from 'rsg-components/Examples';
import Components from 'rsg-components/Components';
import SectionRenderer from 'rsg-components/Section/SectionRenderer';

export default class Section extends Component {
	static propTypes = {
		highlightTheme: PropTypes.string.isRequired,
		section: PropTypes.object.isRequired,
		sidebar: PropTypes.bool,
	};

	renderContent(highlightTheme, examples) {
		if (!examples) {
			return null;
		}

		return (
			<Examples highlightTheme={highlightTheme} examples={examples} />
		);
	}

	renderComponents(highlightTheme, components, sections, sidebar) {
		if (!components && !sections) {
			return null;
		}

		return (
			<Components
				highlightTheme={highlightTheme}
				components={components || []}
				sections={sections || []}
				sidebar={sidebar}
			/>
		);
	}

	render() {
		const { highlightTheme, section, sidebar } = this.props;

		return (
			<SectionRenderer
				name={section.name}
				content={this.renderContent(highlightTheme, section.content)}
				components={this.renderComponents(highlightTheme, section.components, section.sections, sidebar)}
			/>
		);
	}
}
