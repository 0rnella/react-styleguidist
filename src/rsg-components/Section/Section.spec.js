import React from 'react';
import noop from 'lodash/noop';
import Examples from '../Examples';
import Components from '../Components';
import Section from './Section';
import { SectionRenderer } from './SectionRenderer';

const section = {
	name: 'Foo',
	content: [
		{
			type: 'code',
			content: '<button>OK</button>',
			evalInContext: noop,
		},
		{
			type: 'markdown',
			content: 'Hello *world*!',
		},
	],
	components: [],
};

it('should render component renderer', () => {
	const actual = shallow(
		<Section
			section={section}
		/>
	);

	expect(shallowToJson(actual)).toMatchSnapshot();
});

it('render should render component', () => {
	const actual = shallow(
		<SectionRenderer
			classes={{}}
			name={section.name}
			content={<Examples examples={section.content} />}
			components={<Components components={[]} sections={[]} />}
		/>
	);

	expect(shallowToJson(actual)).toMatchSnapshot();
});
