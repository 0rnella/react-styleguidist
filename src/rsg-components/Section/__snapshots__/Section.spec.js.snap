exports[`test render should render component 1`] = `
<div>
  <h2
    id="Foo">
    Foo
  </h2>
  <Examples
    examples={
      Array [
        Object {
          "content": "<button>OK</button>",
          "evalInContext": [Function],
          "type": "code",
        },
        Object {
          "content": "Hello *world*!",
          "type": "markdown",
        },
      ]
    } />
  <Components
    components={Array []}
    sections={Array []} />
</div>
`;

exports[`test should render component renderer 1`] = `
<_class
  components={
    <Components
      components={Array []}
      sections={Array []} />
  }
  content={
    <Examples
      examples={
        Array [
          Object {
            "content": "<button>OK</button>",
            "evalInContext": [Function],
            "type": "code",
          },
          Object {
            "content": "Hello *world*!",
            "type": "markdown",
          },
        ]
      } />
  }
  name="Foo" />
`;
